zkvm/src/vm.rs

```diff
@@ -713,44 +713,31 @@ where
     }
 
     fn call(&mut self) -> Result<(), VMError> {
-        // Pop program, blinding factor, contract, and predicate
-        let prog = self.pop_item()?.to_data()?;
-        let blinding = self.pop_item()?.to_data()?;
+        let program_data = self.pop_item()?.to_data()?;
+        let program = program_data.clone().to_program()?;
+        let call_proof = self.pop_item()?.to_data()?.to_call_proof()?;
         let contract = self.pop_item()?.to_contract()?;
         let predicate = contract.predicate;
 
-        // 0 = -P + h(prog) * B2
+        let signing_key = call_proof.signing_key;
+        let neighbors = call_proof.neighbors;
+
+        // Compute Merkle root M.
+        let transcript = Transcript::new(b"ZkVM.taproot");
```

Comment:
in addition to the above suggestion, let's also move the root computation into the `predicate.prove_taproot`? So this code does not have to assemble the proving, but can use a single function to prove the relation between `program`, `call_proof` and `predicate`.